### PR TITLE
Release 1.0.0

### PR DESCRIPTION
**Pull request by vhrychenko**
- Developer(s): @valerio8787 @vhrychenko

- Peer Reviewer: 

- Tickets:
        https://spryker.atlassian.net/browse/ECO-777

- Academy PR: 

#### Please confirm

- [ ] No new OS components - or they have been approved by the legal department
- [ ] All new or heavily changed classes get an "A" rating (Scrutinizer), except Facades, Factories, QueryContainers and Tests

#### Documentation

- [ ] Functional documentation provided or in progress?
- [ ] Integration guide for projects provided or not needed?
- [ ] Migration guides for all contained majors provided or not needed?

#### Release Table

   Module                | Release Type         | Constraints         |
   :--------------------- | :------------------------ | :--------------------- |
   AkeneoPim               |       major                 |                       |

#### Release Notes
 - 1.0 version for akeneo-pim module.

-----------------------------------------

#### Module AkeneoPim

_**Patch:** Backwards-compatible bug fix_

Def of done (by the responsible developer):
- [ ] All changes are backwards-compatible. The outdated code is marked as deprecated.
- [ ] The change is isolated and not mixed with any cleanups or other changes.
- [ ] New code fits the architecture rules.
- [ ] All new or changed business logic is covered by functional tests (in Zed business layer).

##### Change log
 - 1.0 version for akeneo-pim module.

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -2,7 +2,8 @@
   "name": "spryker-eco/akeneo-pim",
   "description": "Akeneo PIM Integration Module",
   "require": {
-    "akeneo/api-php-client": "^1.0@beta"
+    "akeneo/api-php-client": "^1.0@beta",
+    "php-http/guzzle6-adapter": "*"
```
Comment (gerner-spryker) [inline composer.json]: please clarify version

File: .gitignore [hunk 2]
```diff
@@ -0,0 +1,10 @@
+# IDEs
```
Comment (gerner-spryker) [inline .gitignore:1]: The `config.dist.php` feels much more like an integration/migration guide. Why was it provided in the code base?

File: composer.json [hunk 3]
```diff
@@ -2,7 +2,8 @@
   "name": "spryker-eco/akeneo-pim",
   "description": "Akeneo PIM Integration Module",
   "require": {
```
Comment (gerner-spryker) [inline composer.json:4]: php >=7.1 is missing

File: src/SprykerEco/Shared/AkeneoPim/AkeneoPimConstants.php [hunk 4]
```diff
@@ -13,11 +13,9 @@
  */
```
Comment (gerner-spryker) [inline src/SprykerEco/Shared/AkeneoPim/AkeneoPimConstants.php]: Remove default class comment

File: src/SprykerEco/Shared/AkeneoPim/AkeneoPimConstants.php [hunk 5]
```diff
@@ -13,11 +13,9 @@
...
 class AkeneoPimConstants
 {
-
     const HOST = 'akeneo.host';
```
Comment (gerner-spryker) [inline src/SprykerEco/Shared/AkeneoPim/AkeneoPimConstants.php]: `public` access modifier is missing from constants (php7.1>)

File: src/SprykerEco/Shared/AkeneoPim/AkeneoPimConstants.php [hunk 6]
```diff
@@ -13,11 +13,9 @@
  */
 class AkeneoPimConstants
```
Comment (gerner-spryker) [inline src/SprykerEco/Shared/AkeneoPim/AkeneoPimConstants.php]: This file should be named according spryker naming (https://spryker.atlassian.net/wiki/spaces/RFC/pages/7438355/RFC+Env+Constants+vs+Config+Constants):

AkeneoPimConfig.php

File: src/SprykerEco/Shared/AkeneoPim/AkeneoPimConstants.php [hunk 7]
```diff
@@ -13,11 +13,9 @@
...
 {
-
     const HOST = 'akeneo.host';
     const USERNAME = 'akeneo.username';
     const PASSWORD = 'akeneo.password';
     const CLIENT_ID = 'akeneo.client.id';
     const CLIENT_SECRET = 'akeneo.client.secret';
```
Comment (gerner-spryker) [inline src/SprykerEco/Shared/AkeneoPim/AkeneoPimConstants.php]: This way, the constants are actually closed against extension, please open them up like

```
protected const USERNAME = 'akeneo.username';
...

public function getUsername()
{
   return static::USERNAME;
}
```

Actually this is not the username but the parameter identifier, so also some kind of name change would be necessary like : `getUsernameIdentifier() return static::IDENTIFIER_USERNAME` or something like this.

File: .gitignore (hunk 2 above)
Comment (gerner-spryker) [reply .gitignore:1]: Also `jobs.dist.php`

File: src/SprykerEco/Service/AkeneoPim/AkeneoPimConfig.php [hunk 8]
```diff
@@ -5,18 +5,13 @@
...
+namespace SprykerEco\Service\AkeneoPim;
 
 use Spryker\Zed\Kernel\AbstractBundleConfig;
 use SprykerEco\Shared\AkeneoPim\AkeneoPimConstants;
 
-/**
- * Class AkeneoPimConfig
- * @package SprykerEco\Zed\AkeneoPim
- */
 class AkeneoPimConfig extends AbstractBundleConfig
```
Comment (gerner-spryker) [inline src/SprykerEco/Service/AkeneoPim/AkeneoPimConfig.php:13]: In this class, the shared configs should be used

```
return $this->getSharedConfig()->getUsernameIdentifier();

```

File: src/SprykerEco/Service/AkeneoPim/AkeneoPimService.php [hunk 9]
```diff
@@ -0,0 +1,782 @@
...
+
+use Spryker\Service\Kernel\AbstractService;
+
+/**
+ * @method \SprykerEco\Service\AkeneoPim\AkeneoPimServiceFactory getFactory()
+ */
+class AkeneoPimService extends AbstractService implements AkeneoPimServiceInterface
+{
+    /**
+     * Specification:
```
Comment (gerner-spryker) [inline src/SprykerEco/Service/AkeneoPim/AkeneoPimService.php]: This file should contain only the `{@ineheritdoc}` parts, the specification stays in the interface

File: src/SprykerEco/Service/AkeneoPim/Api/Adapter/AdapterFactory.php [hunk 10]
```diff
@@ -0,0 +1,211 @@
...
+use SprykerEco\Service\AkeneoPim\Api\Adapter\Family\FamilyVariantApiAdapter;
+use SprykerEco\Service\AkeneoPim\Api\Adapter\Family\MeasureFamilyApiAdapter;
+use SprykerEco\Service\AkeneoPim\Api\Adapter\Product\AssociationTypeApiAdapter;
+use SprykerEco\Service\AkeneoPim\Api\Adapter\Product\ProductApiAdapter;
+use SprykerEco\Service\AkeneoPim\Api\Adapter\Product\ProductMediaFileApiAdapter;
+use SprykerEco\Service\AkeneoPim\Api\Adapter\Product\ProductModelApiAdapter;
+use SprykerEco\Service\AkeneoPim\Api\Adapter\Sdk\AkeneoPimSdkFactory;
+use SprykerEco\Service\AkeneoPim\Api\Wrapper\WrapperFactoryInterface;
+
+class AdapterFactory implements AdapterFactoryInterface
```
Comment (gerner-spryker) [inline src/SprykerEco/Service/AkeneoPim/Api/Adapter/AdapterFactory.php]: The AdapterFactory works with Spryker objects already. I don't see the point to have it separately from the normal factory.

File: src/SprykerEco/Service/AkeneoPim/Api/Adapter/Attributes/AttributeApiAdapter.php [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+     * @param \SprykerEco\Service\AkeneoPim\Api\Wrapper\WrapperFactoryInterface $wrapperFactory
+     */
+    public function __construct(AkeneoPimClientInterface $akeneoPimClient, WrapperFactoryInterface $wrapperFactory)
+    {
+        $this->akeneoPimClient = $akeneoPimClient;
+        $this->wrapperFactory = $wrapperFactory;
+    }
+
+    /**
+     * @inheritdoc
```
Comment (gerner-spryker) [inline src/SprykerEco/Service/AkeneoPim/Api/Adapter/Attributes/AttributeApiAdapter.php]:
```
{@inheritdoc}
```


https://manual.phpdoc.org/HTMLSmartyConverter/HandS/phpDocumentor/tutorial_tags.inlineinheritdoc.pkg.html

File: src/SprykerEco/Service/AkeneoPim/Api/Adapter/AbstractApiAdapter.php [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+/**
+ * MIT License
+ * Use of this software requires acceptance of the Evaluation License Agreement. See LICENSE file.
+ */
+
+namespace SprykerEco\Service\AkeneoPim\Api\Adapter;
+
+use Akeneo\Pim\ApiClient\AkeneoPimClient;
+
+abstract class AbstractApiAdapter implements ApiAdapterInterface
```
Comment (gerner-spryker) [inline src/SprykerEco/Service/AkeneoPim/Api/Adapter/AbstractApiAdapter.php]: Shouldn't this class be implemented by the other classes in this directory? Currently there is no usage

File: src/SprykerEco/Service/AkeneoPim/Api/Adapter/AdapterFactory.php [hunk 13]
```diff
@@ -0,0 +1,211 @@
...
+{
+    /**
+     * @var \SprykerEco\Service\AkeneoPim\AkeneoPimConfig
+     */
+    protected $config;
+
+    /**
+     * @var \SprykerEco\Service\AkeneoPim\Api\Wrapper\WrapperFactoryInterface
+     */
+    private $wrapperFactory;
```
Comment (gerner-spryker) [inline src/SprykerEco/Service/AkeneoPim/Api/Adapter/AdapterFactory.php]: why private?

File: src/SprykerEco/Service/AkeneoPim/AkeneoPimService.php [hunk 14]
```diff
@@ -0,0 +1,782 @@
...
+     *  - Gets a cursor to iterate over a list of attributes.
+     *
+     * @api
+     *
+     * @param string $attributeCode Code of the attribute
+     * @param int $pageSize The size of the page returned by the server.
+     * @param array $queryParameters Additional query parameters to pass in the request
+     *
+     * @return \Akeneo\Pim\ApiClient\Pagination\ResourceCursorInterface
```
Comment (gerner-spryker) [inline src/SprykerEco/Service/AkeneoPim/AkeneoPimService.php]: Please fix the return types

File: .gitignore [hunk 15]
```diff
@@ -0,0 +1,10 @@
...
+/.idea
+/.project
+/nbproject
+/.buildpath
+/.settings
+*.sublime-*
+*.AppleDouble
+*.AppleDB
+*.AppleDesktop
```
Comment (gerner-spryker) [inline .gitignore]: please add phpstan level 7 check (phpstan.json)

File: composer.json (hunk 1 above)
Comment (gerner-spryker) [reply composer.json]: kernel is also missing

**Review by gerner-spryker — CHANGES_REQUESTED** [review]
I really liked it how the adapted client is wrapped everywhere, but it still feels a bit mixed in some places. The Akenea namespace is introduced in the middle of the module, it feels a bit wrong. I will try to come up with an actual alternative proposal if i can.

File: src/SprykerEco/Service/AkeneoPim/Api/Adapter/AbstractApiAdapter.php (hunk 12 above)
Comment (vhrychenko) [reply src/SprykerEco/Service/AkeneoPim/Api/Adapter/AbstractApiAdapter.php]: Looks like this abstract class is a redundant class here.

File: src/SprykerEco/Service/AkeneoPim/Api/Adapter/AdapterFactory.php (hunk 13 above)
Comment (vhrychenko) [reply src/SprykerEco/Service/AkeneoPim/Api/Adapter/AdapterFactory.php]: Changed to protected

File: src/SprykerEco/Shared/AkeneoPim/AkeneoPimConstants.php (hunk 5 above)
Comment (vhrychenko) [reply src/SprykerEco/Shared/AkeneoPim/AkeneoPimConstants.php]: Done

File: src/SprykerEco/Shared/AkeneoPim/AkeneoPimConstants.php (hunk 4 above)
Comment (vhrychenko) [reply src/SprykerEco/Shared/AkeneoPim/AkeneoPimConstants.php]: Done

File: composer.json (hunk 3 above)
Comment (vhrychenko) [reply composer.json:4]: Done

File: .gitignore (hunk 2 above)
Comment (vhrychenko) [reply .gitignore:1]: I agree with this notice. Removed.

File: composer.json (hunk 1 above)
Comment (vhrychenko) [reply composer.json]: kernel is added

File: src/SprykerEco/Service/AkeneoPim/AkeneoPimService.php (hunk 14 above)
Comment (vhrychenko) [reply src/SprykerEco/Service/AkeneoPim/AkeneoPimService.php]: What is wrong here?

File: src/SprykerEco/Service/AkeneoPim/Api/Adapter/Attributes/AttributeApiAdapter.php (hunk 11 above)
Comment (vhrychenko) [reply src/SprykerEco/Service/AkeneoPim/Api/Adapter/Attributes/AttributeApiAdapter.php]: Done

File: src/SprykerEco/Service/AkeneoPim/AkeneoPimService.php (hunk 9 above)
Comment (vhrychenko) [reply src/SprykerEco/Service/AkeneoPim/AkeneoPimService.php]: Done

File: src/SprykerEco/Service/AkeneoPim/Api/Adapter/AdapterFactory.php (hunk 10 above)
Comment (vhrychenko) [reply src/SprykerEco/Service/AkeneoPim/Api/Adapter/AdapterFactory.php]: This AdapterFactory divides service object creation logic and adapter object creation logic to separate factories. I think it's ok.

File: .gitignore (hunk 15 above)
Comment (vhrychenko) [reply .gitignore]: Done

File: src/SprykerEco/Shared/AkeneoPim/AkeneoPimConstants.php (hunk 6 above)
Comment (vhrychenko) [reply src/SprykerEco/Shared/AkeneoPim/AkeneoPimConstants.php]: I think you confused. In this class, we have keys which used in the config_local.php and after we use it in AkeneoPimConfig.php like eturn $this->get(AkeneoPimConstants::AKENEO_PIM_HOST);

File: src/SprykerEco/Service/AkeneoPim/AkeneoPimService.php (hunk 14 above)
Comment (gerner-spryker) [reply src/SprykerEco/Service/AkeneoPim/AkeneoPimService.php]: Service returns a 3rd party class, and actually it doesn't :)

File: src/SprykerEco/Shared/AkeneoPim/AkeneoPimConstants.php (hunk 6 above)
Comment (gerner-spryker) [reply src/SprykerEco/Shared/AkeneoPim/AkeneoPimConstants.php]: Then it's worse :) Please check the RFC

File: src/SprykerEco/Service/AkeneoPim/Api/Adapter/AdapterFactory.php (hunk 10 above)
Comment (gerner-spryker) [reply src/SprykerEco/Service/AkeneoPim/Api/Adapter/AdapterFactory.php]: I will evaluate this again, when i check the 3rd party object wrapping.

File: src/SprykerEco/Shared/AkeneoPim/AkeneoPimConstants.php (hunk 6 above)
Comment (vhrychenko) [reply src/SprykerEco/Shared/AkeneoPim/AkeneoPimConstants.php]: Discussed in slack

File: src/SprykerEco/Service/AkeneoPim/AkeneoPimServiceFactory.php [hunk 16]
```diff
@@ -0,0 +1,34 @@
...
+
+/**
+ * @method \SprykerEco\Service\AkeneoPim\AkeneoPimConfig getConfig()
+ */
+class AkeneoPimServiceFactory extends AbstractServiceFactory
+{
+    /**
+     * @return \SprykerEco\Service\AkeneoPim\Api\Adapter\AdapterFactoryInterface
+     */
+    public function createAkeneoPimAdapterFactory()
```
Comment (PhilinTv) [inline src/SprykerEco/Service/AkeneoPim/AkeneoPimServiceFactory.php]: PHP7 Typehints please

File: src/SprykerEco/Service/AkeneoPim/Api/Adapter/Attributes/AttributeApiAdapter.php [hunk 17]
```diff
@@ -0,0 +1,89 @@
...
+        $page = $this->akeneoPimClient
+            ->getAttributeApi()
+            ->listPerPage($limit, $withCount, $queryParameters);
+
+        return $this->wrapperFactory
+            ->createAkeneoPage($page);
+    }
+
+    /**
+     * @inheritdoc
```
Comment (PhilinTv) [inline src/SprykerEco/Service/AkeneoPim/Api/Adapter/Attributes/AttributeApiAdapter.php]: Where are curved brackets?

**Review by PhilinTv — CHANGES_REQUESTED** [review]
1. Injected namespaces should be in dependency folder. 
Please move adapters to Modules/Dependency/External. 

2. Apply PHP7 convention
- const visibility
- typehints

**Review by alex-galych — APPROVED** [review]

File: .scrutinizer.yml [hunk 18]
```diff
@@ -0,0 +1,18 @@
+build:
+    dependencies:
+        before:
+            - composer config repositories.spryker composer https://code.spryker.com/repo/private
```
Comment (dereuromark) [inline .scrutinizer.yml:4]: Please remove outdated repo info.

File: LICENSE [hunk 19]
```diff
@@ -0,0 +1,21 @@
+MIT License
+
+Copyright (c) 2018 SPRYKER SYSTEMS GMBH
```
Comment (dereuromark) [inline LICENSE:3]: This is not the right template.
Please see confluence!

**Pull request merged**